### ISSUE TITLE
Applying themes

### ISSUE THREAD

**Issue by frontierpsycho** [carlmjohnson/tumblr-importr#4]
I am a bit puzzled about how to apply themes to the resulting blogs. I ran the importer successfully, and also realized I probably need to copy over the `layouts/` folder from this repository to my blog (it wasn't done automatically). But then, even though I've installed a theme and specified it in `config.toml`, it's not applied, my blog is plain.

Any idea if I did something wrong, and if not, what the correct method to theme the blog is?

**Comment by earthboundkid** [comment]
Thanks for using this. It sounds like you didn't do anything wrong. Any normal Hugo template that you use with this importer will need to be partially rewritten to work with it. I'm not sure if you saw this "philosophy" section of the README, but it's good background on why:

> When converting a Tumblr blog to Hugo, you may initially think you want all your content converted to Markdown files. For example, you may think you want your link posts to become something like `### Link: [$TITLE]($LINK)↵↵$CONTENT`. The trouble with this approach is that converting to Markdown loses formatting information from Tumblr and locks you into a single representation of the data which cannot be easily changed later.
>
> How tumblr-importr works instead is it reads the common post metadata out of the Tumblr API (title, URL, slug, date, etc.) and writes that in the format Hugo expects, then it makes all of the other data from Tumblr on the post available as a custom parameter. Now you can format your link posts using Hugo's templating language to make it look exactly how you want:
>
> ```html
>  <h3>Link: <a href="{{ .Params.tumblr.url }}">{{ .Params.tumblr.title }}</a></h3>
>
>  {{ .Params.tumblr.description | safeHTML }}
>```
>
>If you decide the `H3` should be an `H2` or the content needs a wrapper `<div class="content">` or you want to change "Link:" to be an emoji 🔗, all you need to do is change your Hugo theme, rather than going back and reformatting all your Markdown files. All of the information that Tumblr had on the post is available, making it possible to fully replicate a Tumblr theme in Hugo without any information loss.

The side effect of this philosophy is that you're going to need to change existing theme to make them work with this data. You can see some of the basics in the layouts directory here, but it's really just a suggested starting point. If I had more time, I would love to have a better sample theme or write some adaptations to show how it works.

If you look at the [_default/single.html](https://github.com/carlmjohnson/tumblr-importr/blob/master/layouts/_default/single.html), the relevant section is here:

```html
        {{ .Render "content" }}
```

That means to render the content.html file according to the current page's type. A normal content.html looks like _defaults/content.html, which calls `{{ .Content }}`. In other words, it's just rendering the page's Markdown as is. This is what normal themes do because it's what you're expected to do by Hugo. What you'll need to do is to go into the theme you want and change `{{ .Content }}` to `{{ .Render "content" }}` and then add different content.html for the different Tumblr page types. So for example, for Tumblr's video page type, I have tumblr-video/content.html, which overrides the normal content.html and looks like this:

```html
{{ range last 1 .Params.tumblr.player }}
    <div class="vidblock">{{ .embed_code | safeHTML }}</div>
{{ end }}

<div class="caption">{{ .Params.tumblr.caption | safeHTML }}</div>
```

In other words, it says "if a page has type tumblr-video, instead of rendering Markdown (which isn't there), look at the .tumblr.play data and render the last embed code in that list, followed by the .tumblr.caption."

The other page types have similar content.html adaptations, such as showing the picture for a photo post or the link for a link post.

Hope that helps. Let me know if any part of that didn't make sense.

**Comment by earthboundkid** [comment]
One more comment: It's possible that you'll find it easier to go from a custom Tumblr theme to a custom Hugo theme than vice versa. E.g. this snippet from Tumblr theme documentation:

```html
        <ol id="posts">
            {block:Posts}{block:Text}
                    <li class="post text">
                        {block:Title}
                            <h3><a href="{Permalink}">{Title}</a></h3>
                        {/block:Title}{Body}
                    </li>
                {/block:Text}{block:Photo}
```

would turn into something like this in Hugo markdown:

```html
        <ol id="posts">
            {{ range .Pages }}{{ if eq .Section "tumblr-text" }}
                    <li class="post text">
                            {{ if eq .Kind "page" }}
                                <h3><a href="{{ .Permalink }}">{{ .Title }}</a></h3>
                            {{ end }}
                        {{ .Params.tumblr.content | safeHTML }}
                    </li>
                {{ end }}{{ if eq .Section "tumblr-photo" }}
```

**Comment by frontierpsycho** [comment]
Thanks for the detailed explanation. I guess I have to delve into Hugo themes then :) 

At the moment I don't understand what to do but I'll try to read and figure it out.

**Comment by frontierpsycho** [comment]
I have one immediate question, though: let's say I manage to change an existing theme to render my tumblr posts the way I want it to. However, when I add new posts, I'm probably not going to follow the structure that the posts imported from tumblr have. So my guess is that any new blog posts I write will not work with the theme I created to accommodate my tumblr posts. Does that sound correct or am I misunderstanding something?

EDIT: Never mind, I think I understand. The tumblr posts have a specific type that is unique to tumblr (tumblr-text, tumblr-video and so on), so they will not clash with my other posts.

**Comment by earthboundkid** [comment]
Yes, your edit is correct. That’s what I did with blog.carlmjohnson.net, where the old posts are Tumblr formatted but the new posts are normal Hugo files.
